core.py

```diff
@@ -577,8 +602,11 @@ def loop(env, minds, vision=Vision(), n_episodes=1, max_steps=-1, policy='determ
     except KeyboardInterrupt:
         # Finish loop when aborted
         log.critical("KeyboardInterrupt, safely terminate loop and exit")
-        callbacks_list.on_loop_finish(True)
+        callbacks_list.on_loop_end(True)
         sys.exit()
 
     # Finish loop as planned
-    callbacks_list.on_loop_finish(False)
+    callbacks_list.on_loop_end(False)
+
+# This import has to be in here because of circular import between 'core' and 'utils'
+from .utils import unpack
```

Comment:
To wygląda bardzo słabo. Cykliczna zależność oznacza że coś jest nie tak. Interfejsy, tzn. klasy abstrakcyjne powinny być wyniesione poza ten plik przykładowo do `interface.py`, a tutaj jako w `core.py` powinniśmy mieć definicje głównych funkcji z humblerl

Reply:
Brzmi nieźle, poprawię.

Reply:
Podzieliłem, zerknij teraz.